scrape/target.go

```diff
@@ -459,47 +459,27 @@ func PopulateLabels(lb *labels.Builder, cfg *config.ScrapeConfig, noDefaultPort
 
 	// addPort checks whether we should add a default port to the address.
 	// If the address is not valid, we don't append a port either.
-	addPort := func(s string) (string, string, bool) {
+	hostPort := func(s string) (string, string) {
 		// If we can split, a port exists and we don't have to add one.
 		if host, port, err := net.SplitHostPort(s); err == nil {
-			return host, port, false
+			return host, port
 		}
-		// If adding a port makes it valid, the previous error
-		// was not due to an invalid address and we can append a port.
-		_, _, err := net.SplitHostPort(s + ":1234")
-		return "", "", err == nil
+
+		return "", ""
 	}
 
 	addr := lb.Get(model.AddressLabel)
 	scheme := lb.Get(model.SchemeLabel)
-	host, port, add := addPort(addr)
-	// If it's an address with no trailing port, infer it based on the used scheme
-	// unless the no-default-scrape-port feature flag is present.
-	if !noDefaultPort && add {
-		// Addresses reaching this point are already wrapped in [] if necessary.
-		switch scheme {
-		case "http", "":
-			addr += ":80"
-		case "https":
-			addr += ":443"
-		default:
-			return labels.EmptyLabels(), labels.EmptyLabels(), fmt.Errorf("invalid scheme: %q", cfg.Scheme)
-		}
-		lb.Set(model.AddressLabel, addr)
-	}
+	host, port := hostPort(addr)
 
-	if noDefaultPort {
-		// If it's an address with a trailing default port and the
-		// no-default-scrape-port flag is present, remove the port.
-		switch port {
-		case "80":
-			if scheme == "http" {
-				lb.Set(model.AddressLabel, host)
-			}
-		case "443":
-			if scheme == "https" {
-				lb.Set(model.AddressLabel, host)
-			}
+	switch port {
+	case "80":
```

Comment:
A comment here might still be good? E.g.
`If it's an address with a trailing port that corresponds to the scheme, remove the the port.`

Reply:
The weird thing is that I couldn't see this behavior in practice. Let me investigate more.

Reply:
Got it now: So what happens is that we remove the default port from the address label, but _not_ from the instance label. I am actually not sure that this was intentional when the feature flag was implemented. It's more an artifact of modifying `addr` or (like in this part of the code) just setting the `model.AddressLabel` directly.

My current gut feeling is that we should _not_ remove the port at all, even if it is the default, or in other words, we really leave it to the user if the port should be there or not.